spaceship/templates/mission.html

```diff
@@ -26,75 +77,284 @@ <h3>{{ mission.title }}</h3>
+<h5>Week {{ week }}</h5>
+</div>
 {% endfor %}
 </section>
 
 <section class="bonus col-sm-6">
 <h4>Secondary Goals</h4>
 <h5>Any time</h5>
-{% for goal in mission.goals: %}
-{% if not goal.primary_for_mission %}
-{% include "goal.html" %}
-{% endif %}
-{% endfor %}
+<div data-week="0">
+  {% for slot in mission.goal_slots(week=0): %}
+  {% set goal = slot.goal %}
+  {% set active = week_of_mission > 0 and week_of_mission <= mission.duration_in_weeks %}
+  {% include "goal.html" %}
+  {% endfor %}
+</div>
 </section>
 </div>
+
+{% if can_edit %}
+<div class="row">
+<section class="unscheduled col-sm-6">
+<h4>Unscheduled Goals</h4>
+<button type="button" id="search" class="btn btn-primary">Search</button>
+<div data-week="-1">
+</div>
+</section>
 </div>
+{% endif %}
 
 {% endblock %}
 
 {% block scripts %}
 {{ super() }}
+
+{% if can_edit %}
+<script>
+(function() {
+var state = {{ mission_js_model|tojson }};
+var focusedGoalId = -1;
+render();
+
+function render() {
+  $('.goal').detach().appendTo($('[data-week="-1"]'));
+  var schedule = state['schedule'];
+  for (var week = 0; week < {{ max_weeks }}; week++) {
+    var weekDiv = $('[data-week="' + week + '"]');
+    if (week >= schedule.length) {
+      $(weekDiv).hide();
+      continue;
+    }
+    $(weekDiv).show();
+    for (var i = 0; i < schedule[week].length; i++) {
+      var slot = schedule[week][i];
+      var goalId = slot['goal_id'];
+      var goal = $('.goal[data-goal="' + goalId + '"]');
+      appendEditControls(goal);
+      weekDiv.append(goal);
+    }
+  }
+  $('#duration').val('' + (schedule.length - 1));
+  $('.goal').each(function() {
+    var goalId = parseInt($(this).data('goal'));
+    $(this).toggleClass('focused-goal', focusedGoalId == goalId);
+  });
+}
+
+function appendEditControls(goal) {
+  if (goal.find('.edit-controls').length !== 0) {
+    return;
+  }
+  var dom = goal.append(`<div class="edit-controls btn-toolbar" role="toolbar" aria-label="Scheduling toolbar">
+    <div class="btn-group mr-2" role="group" aria-label="Move goals">
+      <button type="button" class="move-goal-down btn btn-primary" title="Move down"><i class="fas fa-arrow-circle-down"></i></button>
+      <button type="button" class="move-goal-up btn btn-primary" title="Move up"><i class="fas fa-arrow-circle-up"></i></button>
+    </div>
+    <div class="btn-group mr-2 ml-auto" role="group" aria-label="Change importance">
+      <button type="button" class="make-goal-primary btn btn-primary ml-auto" title="Make primary"><i class="fas fa-calendar-plus"></i></button>
+      <button type="button" class="make-goal-secondary btn btn-primary ml-auto" title="Make secondary"><i class="fas fa-calendar-minus"></i></button>
+    </div>
+    <div class="btn-group mr-2" role="group" aria-label="Remove">
+      <button type="button" class="remove-goal btn btn-primary" title="Remove"><i class="fas fa-times-circle"></i></button>
+    </div>
+  </div>`);
+  $(dom).find('.remove-goal').click(function() {
```

Comment:
we should really start using something like `vuejs` to make a lot of this dom manipulation declarative

Reply:
Hrm looks interesting, I will look into it. It is seeming like we are going to have a fair bit of js here after all and jquery is feeling awfully crusty...